### PR TITLE
feat: version-aware error filtering + opt-in bot identity (#401)

### LINKED ISSUE #401: Feature: Version-aware issue resolution + opt-in bot identity in error reporting
### 1. Version-aware resolution
When an issue is resolved (e.g. LLM timeout fixed in v2026.3.112), old versions (v2026.3.110) still running will re-report the same error, causing GlitchTip to reopen the issue. GlitchTip doesn't support Sentry's `resolve in release` feature.

**Desired behavior:** Errors from versions older than the `fixedInRelease` version should be silently dropped via `beforeSend`, so resolved issues stay resolved until a *newer* version reproduces them.

### 2. Opt-in bot identity
`botName` tag support exists in error-reporter but:
- Should be **opt-in** (not sent by default) — privacy-first
- Useful for multi-agent setups (Maeve vs Doris) to filter errors by agent
- Config: `errorReporting.botName` already exists, just needs to be explicitly opt-in documented

### beforeSend version filtering
- Add a `resolvedIssues` map in error-reporter config: `{ fingerprint: fixedInVersion }`
- In `beforeSend`, compare event release version against the fixedInVersion
- If event version < fixedInVersion → drop the event (return null)
- If event version >= fixedInVersion → let it through (real regression)

### Bot identity
- Ensure `botName` tag is only sent when explicitly configured
- Document as opt-in in README/config docs
- Add to config schema with clear privacy note

## Context
- GlitchTip at 192.168.1.99:8000 (self-hosted)
- Current release format: `openclaw-hybrid-memory@2026.3.NNN`
- Release is in event tags but not top-level `release` field (may need fix)
- `statusDetails.inRelease` API tested — GlitchTip ignores it

### PR DESCRIPTION
**Pull request by markus-lassfolk**
## Summary

Implements version-aware error filtering and clarifies opt-in bot identity for GlitchTip/Sentry error reporting.

### Changes

**Version-aware `beforeSend` filtering:**
- New `resolvedIssues` config: map of `{ errorFingerprint: fixedInVersion }`
- `beforeSend` compares event release version against fixedInVersion
- Events from older versions are silently dropped (not a regression)
- Events from equal/newer versions pass through (real regression)
- Safe defaults: unparseable versions always pass through

**Utilities:**
- `extractVersion(release)` — parses `openclaw-hybrid-memory@2026.3.110` → `2026.3.110`
- `compareVersions(a, b)` — numeric segment comparison

**Opt-in bot identity:**
- JSDoc updated: `botName`/`botId` explicitly marked as opt-in
- Init logging shows whether bot name is set or omitted (privacy default)

### Tests
19 new tests covering version parsing, comparison, drop/pass-through logic, edge cases, and bot identity opt-in behavior.

**3251 tests pass, `tsc --noEmit` clean.**

Closes #401

---

> [!NOTE]
> **Medium Risk**
> Adds version-based suppression in Sentry `beforeSend`, which could unintentionally drop real errors if fingerprints/versions are misconfigured. Changes are contained to error-reporting config/parsing and have strong test coverage, but affect observability.
> 
> **Overview**
> Adds a new `errorReporting.resolvedIssues` config that maps error fingerprints to the plugin version where they were fixed, and uses it in `beforeSend` to **silently drop** events coming from *older releases* (while allowing equal/newer versions through as regressions).
> 
> Introduces helper utilities (`extractVersion`, `compareVersions`, `shouldDropForResolvedIssue`), wires the new config through parsing/types and `plugin-service` initialization, and clarifies *opt-in* `botId`/`botName` behavior with debug logging. Expands `error-reporter` tests substantially to cover version parsing/comparison, drop/pass-through edge cases, and bot identity logging/sanitization.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 04f7a9a63b757a6cde10009089ad65faeea59b63. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: extensions/memory-hybrid/services/error-reporter.ts [hunk 1]
```diff
@@ -31,15 +32,98 @@ export interface ErrorReporterConfig {
...
+}
+
+/**
+ * Compare two version strings numerically (YYYY.M.N format).
+ * Returns:
+ *   -1 if a < b
+ *    0 if a === b
+ *    1 if a > b
+ */
+export function compareVersions(a: string, b: string): number {
```
Comment (markus-lassfolk) [inline extensions/memory-hybrid/services/error-reporter.ts]: ### 🟡 Dead Code — `compareVersions` not used in production

**Warning**

`compareVersions` is exported and has 6 dedicated tests in the *Version Helpers* suite, but `shouldDropForResolvedIssue` at line 124 calls `isVersionAtLeast` from `../utils/version-check.js` — not this function. Two independent version-comparison implementations now exist and can silently drift.

The *Version Helpers* tests give false confidence: they exercise this dead function, not the comparison path that actually gates event dropping. If `isVersionAtLeast` behaviour ever changes, these tests won't catch it.

**Options:**
1. Remove `compareVersions` and have the tests import and exercise `isVersionAtLeast` + `parseVersion` from `version-check` directly.
2. Or, replace the `isVersionAtLeast` call at line 124 with `compareVersions(eventVersion, fixedInVersion) < 0` and drop the separate import — but note the safe-default semantics differ slightly: `compareVersions` returns `0` for unparseable inputs; `isVersionAtLeast` returns `true`. Both produce "don't drop" in context, so observable behaviour is the same.

File: extensions/memory-hybrid/services/error-reporter.ts [hunk 2]
```diff
@@ -31,15 +32,98 @@ export interface ErrorReporterConfig {
...
+export function shouldDropForResolvedIssue(
+  event: SentryType.Event,
+  resolvedIssues: Record<string, string>,
+  fallbackRelease?: string,
+): boolean {
+  if (!resolvedIssues || Object.keys(resolvedIssues).length === 0) return false;
+
+  const errType = event.exception?.values?.[0]?.type || "Error";
+  const errValue = event.exception?.values?.[0]?.value || "";
+  const fingerprint = `${errType}:${errValue.slice(0, 100)}`;
```
Comment (markus-lassfolk) [inline extensions/memory-hybrid/services/error-reporter.ts]: ### 🟡 Fingerprints match *post-sanitize* values — undocumented operator footgun

**Warning**

`shouldDropForResolvedIssue` is called with the **sanitized** event (after `sanitizeEvent` / `scrubString` runs). So `errValue` here is the scrubbed message, not the raw one.

An operator configuring `resolvedIssues` will likely copy error text from GlitchTip, which shows the *raw* message before scrubbing. If `scrubString` transforms that message (e.g. replaces IPs with `[IP]`, paths with `[PATH]`), the configured key will never match and the filter silently does nothing — with no indication it's broken.

**Recommendation:** Add a comment here noting that `errValue` comes from the sanitized event, and update the `resolvedIssues` JSDoc in both the type and config parser to say:
> Fingerprint keys must match the **post-sanitize** error message (as shown in GlitchTip reports), not the raw exception text.

Also worth noting: this custom `ErrorType:value` fingerprint is distinct from Sentry's own `event.fingerprint` grouping field — operators familiar with Sentry may confuse the two.

File: extensions/memory-hybrid/config/parsers/features.ts [hunk 3]
```diff
@@ -373,6 +373,13 @@ export function parseErrorReportingConfig(cfg: Record<string, unknown>): ErrorRe
   const botNameRaw = typeof errorReportingRaw.botName === "string" ? errorReportingRaw.botName.trim() : "";
   const botName = botNameRaw.length > 0 ? botNameRaw.slice(0, 64) : undefined;
 
+  // Optional resolvedIssues map for version-aware filtering
+  const resolvedIssues = errorReportingRaw.resolvedIssues && typeof errorReportingRaw.resolvedIssues === "object" && !Array.isArray(errorReportingRaw.resolvedIssues)
+    ? Object.fromEntries(
+        Object.entries(errorReportingRaw.resolvedIssues).filter(([_, v]) => typeof v === "string")
```
Comment (markus-lassfolk) [inline extensions/memory-hybrid/config/parsers/features.ts]: ### 🟢 Missing type predicate — type safety is illusory here

**Suggestion**

`Array.filter(([_, v]) => typeof v === 'string')` does not narrow the return type to `[string, string][]` in TypeScript without an explicit type predicate. TypeScript infers `Object.entries(obj: object)` as `[string, any][]`, so the filtered result is still `[string, any][]` and `Object.fromEntries` yields `Record<string, any>`. The assignment to `Record<string, string>` compiles silently because `any` is assignable to `string`.

This passes `tsc --noEmit` today, but the type guarantee is hollow — a non-string that somehow slips through would flow into `resolvedIssues` undetected.

**Fix:**
```typescript
.filter((entry): entry is [string, string] => typeof entry[1] === 'string')
```
This gives a genuine `[string, string][]` and makes `Object.fromEntries` infer `Record<string, string>` correctly.

File: extensions/memory-hybrid/services/error-reporter.ts [hunk 4]
```diff
@@ -31,15 +31,85 @@ export interface ErrorReporterConfig {
...
+
+  const fixedInVersion = resolvedIssues[fingerprint];
+  if (!fixedInVersion) return false;
+
+  const releaseStr = event.release || fallbackRelease || "";
+  const eventVersion = extractVersion(releaseStr);
+  if (!eventVersion) return false; // Can't parse → safe default: let through
+
+  return compareVersions(eventVersion, fixedInVersion) < 0;
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/services/error-reporter.ts]: Fixed in commit 04f7a9a. `shouldDropForResolvedIssue` now validates `fixedInVersion` with `/^\d+\.\d+\.\d+/.test(fixedInVersion)` before calling `compareVersions`. Malformed values short-circuit with `false` (safe default: do not drop).

File: extensions/memory-hybrid/services/error-reporter.ts (hunk 2 above)
Comment (markus-lassfolk) [reply extensions/memory-hybrid/services/error-reporter.ts]: Fixed in commit 04f7a9a. Added a JSDoc block to `shouldDropForResolvedIssue` explicitly noting: "NOTE: `errValue` is read from the event and passed through `scrubString()` before building the fingerprint. When called from `beforeSend`, the event has already been through `sanitizeEvent()` (which also applies `scrubString()`), so `resolvedIssues` keys must use the post-sanitize (scrubbed) form of the error message."

**Comment by markus-lassfolk** [comment]
## 📝 Documentation Review — PR #402

### Suggested CHANGELOG entry

Add under `## [Unreleased]` → `### Added`:

```markdown
### Added

- **Version-aware error filtering (`resolvedIssues`) (#401):** New `errorReporting.resolvedIssues` config: a map of `{ errorFingerprint: fixedInVersion }`. The `beforeSend` hook compares the event's release version against `fixedInVersion` — events from older versions are silently dropped (known fix, not a regression), while events from equal or newer versions pass through (potential regression). Safe defaults: unparseable or missing versions always pass through (never drops unknown). Includes `extractVersion()` and `compareVersions()` utilities.
- **Opt-in bot identity for error reporting (`botIdentity`) (#401):** `botName` and `botId` in `errorReporting` config are now explicitly documented as opt-in. Init logging shows whether bot name is set or omitted (privacy-preserving default). Useful for GlitchTip/Sentry release tagging in multi-bot deployments.
```

### CONFIGURATION.md updates needed

The **Error reporting (errorReporting)** section (line ~1577) needs two additions:

**1. Add `resolvedIssues` to the config example and table:**

Update the JSON example to include:

```json
"errorReporting": {
  "enabled": true,
  "consent": true,
  "mode": "community",
  "sampleRate": 1.0,
  "environment": "production",
  "botName": "Maeve",
  "resolvedIssues": {
    "embedding-provider-401": "2026.3.120",
    "unconfigured-provider-chat": "2026.3.110"
  }
}
```

**2. Add rows to the config table:**

| Key | Default | Description |
|-----|---------|-------------|
| `resolvedIssues` | `{}` | Map of `{ errorFingerprint: fixedInVersion }`. Events matching a fingerprint from a release older than `fixedInVersion` are silently dropped by `beforeSend` (known fix). Events from equal/newer versions pass through as potential regressions. Unparseable versions always pass through. |
| `botName` | *(unset)* | **Opt-in.** Friendly name tag for grouping errors by bot in GlitchTip/Sentry. Omitted by default for privacy. |
| `botId` | *(unset)* | **Opt-in.** UUID tag for grouping errors by bot. Omitted by default for privacy. |

> Note: `botName` and `botId` already exist in the table but should have "**Opt-in.**" prepended to their descriptions to match the updated JSDoc and init logging behaviour from this PR.

### README update

No README changes needed — the error reporting section in README already links to CONFIGURATION.md and ERROR-REPORTING.md for details.

---
*Scholar docs review — automated*
